File: VB/S134179/CaptionBar.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Skins
Imports DevExpress.LookAndFeel
Imports DevExpress.XtraEditors
Imports DevExpress.Utils.Drawing

Namespace S134179
	Public Class CaptionBar
		Inherits GroupControl
		Public Sub New()
			MyBase.New()
		End Sub

		Protected Overrides Function CreatePainter() As ObjectPainter
			Select Case LookAndFeel.ActiveStyle
				Case ActiveLookAndFeelStyle.Office2003
					Return New CaptionBarObjectPainter(Me)
				Case ActiveLookAndFeelStyle.WindowsXP
					Return New WindowsXPCaptionBarObjectPainter(Me)
				Case ActiveLookAndFeelStyle.Skin
					Return New SkinCaptionBarObjectPainter(Me, LookAndFeel)
			End Select
			Return New FlatCaptionBarObjectPainter(Me)
		End Function
	End Class

	Public Class SkinCaptionBarObjectPainter
		Inherits SkinGroupObjectPainter
		Public Sub New(ByVal owner As IPanelControlOwner, ByVal provider As ISkinProvider)
			MyBase.New(owner, provider)
		End Sub

		Protected Overrides Function CalcCaptionContentHeight(ByVal info As GroupObjectInfoArgs) As Integer
			Return info.Bounds.Height
		End Function

		Protected Overrides Sub CalcCaptionNew(ByVal info As GroupObjectInfoArgs)
			info.CaptionBounds = info.Bounds
			CalcCaptionElements(info, info.Bounds)
		End Sub
	End Class

	Public Class CaptionBarObjectPainter
		Inherits GroupObjectPainter
		Public Sub New(ByVal owner As IPanelControlOwner)
			MyBase.New(owner)
		End Sub

		Protected Overrides Function CalcCaptionContentHeight(ByVal info As GroupObjectInfoArgs) As Integer
			Return info.Bounds.Height
		End Function

		Protected Overrides Sub CalcCaptionNew(ByVal info As GroupObjectInfoArgs)
			info.CaptionBounds = info.Bounds
			CalcCaptionElements(info, info.Bounds)
		End Sub
	End Class

	Public Class WindowsXPCaptionBarObjectPainter
		Inherits WindowsXPGroupObjectPainter
		Public Sub New(ByVal owner As IPanelControlOwner)
			MyBase.New(owner)
		End Sub

		Protected Overrides Function CalcCaptionContentHeight(ByVal info As GroupObjectInfoArgs) As Integer
			Return info.Bounds.Height
		End Function

		Protected Overrides Sub CalcCaptionNew(ByVal info As GroupObjectInfoArgs)
			info.CaptionBounds = info.Bounds
			CalcCaptionElements(info, info.Bounds)
		End Sub
	End Class

	Public Class FlatCaptionBarObjectPainter
		Inherits FlatGroupObjectPainter
		Public Sub New(ByVal owner As IPanelControlOwner)
			MyBase.New(owner)
		End Sub

		Protected Overrides Function CalcCaptionContentHeight(ByVal info As GroupObjectInfoArgs) As Integer
			Return info.Bounds.Height
		End Function

		Protected Overrides Sub CalcCaptionNew(ByVal info As GroupObjectInfoArgs)
			info.CaptionBounds = info.Bounds
			CalcCaptionElements(info, info.Bounds)
		End Sub
	End Class
End Namespace

File: CS/S134179/CaptionBar.cs
using System;
using DevExpress.Skins;
using DevExpress.LookAndFeel;
using DevExpress.XtraEditors;
using DevExpress.Utils.Drawing;

namespace S134179 {
    public class CaptionBar:GroupControl {
        public CaptionBar() : base() { }

        protected override ObjectPainter CreatePainter() {
            switch (LookAndFeel.ActiveStyle) {
                case ActiveLookAndFeelStyle.Office2003: return new CaptionBarObjectPainter(this);
                case ActiveLookAndFeelStyle.WindowsXP: return new WindowsXPCaptionBarObjectPainter(this);
                case ActiveLookAndFeelStyle.Skin: 
                    return new SkinCaptionBarObjectPainter(this, LookAndFeel);
            }
            return new FlatCaptionBarObjectPainter(this);
        }
    }

    public class SkinCaptionBarObjectPainter :SkinGroupObjectPainter {
        public SkinCaptionBarObjectPainter(IPanelControlOwner owner, ISkinProvider provider) 
            : base(owner, provider) { }

        protected override int CalcCaptionContentHeight(GroupObjectInfoArgs info) {
            return info.Bounds.Height;
        }

        protected override void CalcCaptionNew(GroupObjectInfoArgs info) {
            info.CaptionBounds = info.Bounds;
            CalcCaptionElements(info, info.Bounds);
        }
    }

    public class CaptionBarObjectPainter :GroupObjectPainter {
        public CaptionBarObjectPainter(IPanelControlOwner owner) : base(owner) { }

        protected override int CalcCaptionContentHeight(GroupObjectInfoArgs info) {
            return info.Bounds.Height;
        }

        protected override void CalcCaptionNew(GroupObjectInfoArgs info) {
            info.CaptionBounds = info.Bounds;
            CalcCaptionElements(info, info.Bounds);
        }
    }

    public class WindowsXPCaptionBarObjectPainter :WindowsXPGroupObjectPainter {
        public WindowsXPCaptionBarObjectPainter(IPanelControlOwner owner) : base(owner) { }

        protected override int CalcCaptionContentHeight(GroupObjectInfoArgs info) {
            return info.Bounds.Height;
        }

        protected override void CalcCaptionNew(GroupObjectInfoArgs info) {
            info.CaptionBounds = info.Bounds;
            CalcCaptionElements(info, info.Bounds);
        }
    }

    public class FlatCaptionBarObjectPainter :FlatGroupObjectPainter {
        public FlatCaptionBarObjectPainter(IPanelControlOwner owner) : base(owner) { }

        protected override int CalcCaptionContentHeight(GroupObjectInfoArgs info) {
            return info.Bounds.Height;
        }

        protected override void CalcCaptionNew(GroupObjectInfoArgs info) {
            info.CaptionBounds = info.Bounds;
            CalcCaptionElements(info, info.Bounds);
        }
    }
}
